crypto/params.c

```diff
@@ -0,0 +1,402 @@
+}
+
+static CRYPTO_ONCE params_init_flag = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_params_init)
+{
+    allocated_params = ossl_sa_void_new();
+    return allocated_params != NULL && OPENSSL_atexit(&ossl_params_cleanup);
+}
+
+static void ossl_param_register_alloc(void *string)
+{
+    if (!RUN_ONCE(&params_init_flag, do_params_init))
+        return;
+
+    /* Magic to go here */
+    ossl_sa_void_set(allocated_params, (ossl_uintmax_t)string, string);
+}
+
+void OSSL_PARAM_string_free(void *string)
+{
+    if (!RUN_ONCE(&params_init_flag, do_params_init))
+        return;
+
+    /* Magic to go here */
+    if (ossl_sa_void_get(allocated_params, (ossl_uintmax_t)string) == NULL)
+        return;
+    ossl_sa_void_set(allocated_params, (ossl_uintmax_t)string, NULL);
+    free(string);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, char *buf,
+                                            size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING, buf, bsize,
+                                rsize);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string_ptr(const char *key, char *buf,
+                                                size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING_PTR, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_utf8_string(const OSSL_PARAM *p, char **val, size_t max_len)
+{
+    char *q;
+
+    if (val == NULL || p == NULL)
+        return 0;
+    if (p->data_type != OSSL_PARAM_UTF8_STRING &&
+        p->data_type != OSSL_PARAM_UTF8_STRING_PTR)
+        return 0;
+
+    q = *val;
+    if (q == NULL) {
+        if (p->data_type == OSSL_PARAM_UTF8_STRING_PTR) {
+            *val = *(char **)p->buffer;
+            return 1;
+        }
+        if ((q = OPENSSL_malloc(p->buffer_size)) == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, p->buffer_size);
+        ossl_param_register_alloc(q);
+        return 1;
+    }
+    if (max_len < p->buffer_size)
+        return 0;
+    memcpy(q, p->buffer, p->buffer_size);
+    return 1;
+}
+
+int OSSL_PARAM_set_utf8_string(OSSL_PARAM *p, char *val)
+{
+    if (p == NULL)
+        return 0;
+
+    if (val != NULL) {
+        const size_t len = strlen(val) + 1;
+
+        SET_RETURN_SIZE(p, len);
+        switch (p->data_type) {
+        case OSSL_PARAM_UTF8_STRING:
+            if (p->buffer_size >= len) {
+                memcpy(p->buffer, val, len);
+                return 1;
+            }
+            break;
+
+        case OSSL_PARAM_UTF8_STRING_PTR:
+            *(char **)p->buffer = val;
+            return 1;
+        }
+    } else
+        SET_RETURN_SIZE(p, 0);
+    return 0;
+}
+
+OSSL_PARAM OSSL_PARAM_construct_octet_string(const char *key, void *buf,
+                                             size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING, buf, bsize,
```

Comment:
UTF8 ?

Reply:
Yes, there is a worse problem though :(

Reply:
Oh?

Reply:
Well, was a worse problem.  The buffer for the PTR one below should have been declared void**.  Similarly for utf8.

Fixed now.